lecture2gether_flask/meta_data_provider.py

```diff
@@ -150,3 +157,55 @@ def youtube_video_id_from_url(video_url):
     return None
 
 
+class GoogleDriveMetaDataProvider(MetaDataProvider):
+    def __init__(self, share_link):
+        super().__init__(share_link)
+        self._file_id = drive_file_id_from_share_url(share_link)
+
+        if not self._file_id:
+            raise VideoNotFoundException
+
+        # Check API Key
+        if 'GOOGLE_API_KEY' not in os.environ:
+            raise APIUnauthorizedException
+
+        # Init Google API
+        try:
+            self._drive = googleapiclient.discovery.build(
+                'drive', 'v3', developerKey=os.environ['GOOGLE_API_KEY'], cache_discovery=False)
+        except HttpError as e:
+            if e.resp.status in [400, 401, 403]:
+                raise APIUnauthorizedException
+            else:
+                raise e
+
+    def get_meta_data(self):
+        # Get meta data for file
+        try:
+            file_meta = self._drive.files().get(fileId=self._file_id).execute()
+        except HttpError as e:
+            if e.resp.status in [400, 401, 403]:
+                raise VideoUnauthorizedException
+            elif e.resp.status in [404,]:
+                raise VideoNotFoundException
+            else:
+                raise e
+        self.video_meta_data["title"] = file_meta["name"]
+        self.video_meta_data["mimeType"] = file_meta["mimeType"]
+
+        # Get stream url from API
+        file_download_url = json.loads(self._drive.files().get_media(fileId=self._file_id).to_json())['uri']
+        self.video_meta_data["streamUrl"] = file_download_url
+
+        return super().get_meta_data()
+
+
+def drive_file_id_from_share_url(share_url):
+    """
+    Converts a google drive share link into a google drive file id
+    """
+    url = urlparse(share_url)
+    if url.hostname in ['drive.google.com',]:
+        if re.fullmatch(r'/file/d/[a-zA-Z0-9-_]+/view', url.path):
+            return url.path[len('/file/d/'):-len("/view")]
+    return None
```

Comment:
I get, that this could be a static method, since it does not use any class objects, but I think, this is so specific, that we should move this into the `GoogleDriveMetaDataProvider` class.

Reply:
Same for the yt stuff. I thought about that but used the same layout as Timon.

Reply:
Oh, okay. So maybe, we should indent both?

Reply:
I made both a static method using the decorator.

Reply:
Thanks